lib/TPP/ToBlockLayoutAndBack.cpp

```diff
@@ -557,6 +557,62 @@ struct PropagateThroughPadOp : public OpRewritePattern<tensor::PadOp> {
   }
 };
 
+//===----------------------------------------------------------------------===//
+// BubbleUpThroughFillOp
+//===----------------------------------------------------------------------===//
+
+// Attempt to avoid packing a fill op. Instead create a 'packed' fill.
+// %0 = tensor.empty
+// %packed = tensor.empty
+// %1 = linalg.fill ins(%cst) outs(%0)
+// %2 = tensor.pack %1 into %packed
+// %3 = some_packed_op %2
+// %4 = tensor.unpack %3 into %1
+// --->
+// %0 = tensor.empty
+// %1 = linalg.fill ins(%cst) outs (%packed)
+// %2 = some_packed_op %1
+// %3 = tensor.unpack %2 into %0
+//
+struct BubbleUpThroughFillOp : public OpRewritePattern<tensor::PackOp> {
+  using OpRewritePattern<tensor::PackOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::PackOp packOp,
+                                PatternRewriter &rewriter) const override {
+    Value source = packOp.getSource();
+    auto fillOp = source.getDefiningOp<linalg::FillOp>();
+    if (!fillOp)
+      return failure();
+
+    Value fillRes = fillOp.getResult(0);
+    auto users = fillRes.getUsers();
+
+    // Only two users: a pack and an unpack.
+    if (std::distance(users.begin(), users.end()) != 2)
+      return failure();
+
+    bool foundPack, foundUnPack = false;
+    for (Operation *op : users) {
+      if (isa_and_nonnull<tensor::UnPackOp>(op))
+        foundUnPack = true;
+      if (isa_and_nonnull<tensor::PackOp>(op))
```

Comment:
Why would the user of the Op's result be a `pack`? Or rather, if it's actually a `pack`, isn't that the next relayout operation and not this one?

Reply:
The users of the fill must be a pack or unpack for this patter to apply.

Reply:
Does this answer your question?

Reply:
I understood the code and its intention clearly, I just don't get why it would be a `pack`.

Reply:
Or better still, in which case it would be an `unpack` and why don't you have a test for it?

Reply:
Discussed offline, this looks good!